### ISSUE TITLE
[bug]: Backups don't seem to backup region files...

### ISSUE THREAD

**Issue by ztk4** [FTBTeam/FTB-Mods-Issues#20]
AFAICT the last several backups in my backup folder all have the same block data (region files).
In each backup, my inventory and player position is different, but the world always looks the same.

For some more context, I was working on my base for ~20 minutes when the server came down unexpectedly.
I had 3 backups from this time period, but all 3 of them seem to exclude all of the edits made to the world during this work (specifically, swapping out the floor for a different block, digging a basement, and putting in blocks for the floors, ceiling, and walls of the basement).

In fact, in some of the backups my player's position is in the basement, but since the basement is not cleared out in these backups my player immediately suffocates and dies. 

It's worth noting that some edits to the world (block placements) were included in the backup that occurred after the server had started up. So it seems like some edits were making it to the backups. But the most recent edits weren't.

I was also in the same location for this last ~20 minutes of work, so is it possible that updated versions of loaded chunks aren't being written out? I.e. perhaps the source of chunk data ftb-backups is writing to the backup file is only updated when a chunk is unloaded?

**Mod & Pack version**

Enigmatica 6 v0.5.11
ftb-backups-2.1.1.6

**Forge / Fabric version, if applicable**

FML 1.16.5-36.2.2

**Modpack (with version), if applicable**

Enigmatica 6 v0.5.11

**How to reproduce the issue**

* Set the backups to make a backup every 5 minutes or so (or faster if your computer compresses quickly enough).
* Play around in a world (placing and removing blocks) for 10-20 minutes, but also move your player around and change your inventory around. I'd recommend staying exclusively in just a few chunks as I had done.
* See if the backups contain edits accordingly or if, as in my case, all the backups contain up-to-date player data but identical region data.

If the above reproduction works, it might be interesting to try the same thing, but switching between dimensions and such throughout. This should force the server to write the chunks out to file as they unload, which might force ftb-backups to see newer edits. Not sure.

**Comment by ztk4** [comment]
I have not confirmed if this issue recurs, but as tahg was saying, theoretically a server crash at time X should not affect the integrity of backups for 1-2 hours leading up to time X. I would've expected those backups to be complete s.t. when the server crashes and corrupts the active world on disk, the backups are a viable recovery option.

W.r.t. your last comment, I'm not sure I understand your point. Are you saying that before a backup FTB Backups does _not_ flush chunks, so in general a backup will only backup the chunk info from disk but not flush loaded/buffered chunks in RAM? If so, is there a specific reason not to do this? As in, is it _very_ expensive to do so?

I'm also not sure what you mean by auto save. Is there some other configuration for my server that can force the server to flush chunks to disk periodically? Or is this some mod I would have to add to the server instance?

**Comment by Raidobw2** [comment]
I've only glanced at it last time, but effectively I did not see a line where backups would flush to disk. They only flushed playerdata, but not chunks. Meaning the next backup would use the disk info yes, from a previous save: if I understood this right. No clue why it's coded like that, but as I said on my previous answer, our modder will choose if he wants to change this. Yes it probably would be expensive to save and backup right away. This might be only a backup tool and not a save tool and it would still make a lot of sense.

Here were the lines I looked at if you want to explore it: <https://github.com/FTBTeam/FTB-Backups/blob/3dae1451efd112da840d26602b87f2369014bf00/src/main/java/com/feed_the_beast/mods/ftbbackups/Backups.java#L174>
![image](https://user-images.githubusercontent.com/17010632/137608797-7b025267-bb16-40b6-aa20-21542f86b809.png)


As for autosave, usually on online hosts that's something they already take care of. Basically you can have a task write in the server console every 5 mins for example the command `save-all` and it will save your world periodically. You can do that locally if you host the server too, it just requires scripting and OS knowledge. On Linux you could use a cronjob with the screen command. On Windows or MacOS I never tried it, but it might be more advanced too since the commands could be different. Anyways here is not the place to discuss about that.
<https://forum.feed-the-beast.com/threads/minecraft-custom-linux-backups-using-crontab-and-screen.45575/>

Here is what autosave looks like when using an online host for example:
![image](https://user-images.githubusercontent.com/17010632/137609080-5a86f9a4-24c0-423a-9a21-d6615d034079.png)
On a local server you would only get the last two lines since it's not as complex as the wrapper of the online host (presence of it seen on first line).

**Comment by ztk4** [comment]
Responding here because this just happened to me again. I blue-screened while hosting a server -- when I finally got the server back and running the state on disk _and_ the most recent backup had region data from over an hour prior to the crash. This affected me and a fried who were playing on the server at the time of the BSOD.

Annoyingly, since our playerdata is consistent with our inventories at the time of the last backup, but the region data is stale by over an our, both of our players have "lost" several items: I had several items in my player's inventory at the time of the last flush of region data to disk that I had since placed into chests in the world. Since those region-file edits (the new items in the chest inventories) are effectively lost (only ever existing in RAM prior to the crash), those items are now both **not** in the world _and_ **not** in my inventory...

@Raidobw2 I appreciate the code pointers in your last reply, and agree that the code looks as it behaves: flushes only playerdata and not region data to disk when running backups. IMO `/save-all` should really be part of any sensible backup routine... it just doesn't make sense to me to flush partial state to disk before backing up rather than flushing the entire state to disk. It sounds like the decision for whether this feature belongs in this mod would be to have your modder TAL and decide if they agree and would like to add it. Is there some hotlist I can put this issue on to have it be seen? What are the steps towards realizing a fix here, if that is at all possible?

### Efficiency 

Consider that as you said, most "online hosts ... already take care of [auto /save-all]". If it was terribly expensive to run `/save-all` repeatedly they would already be running into problems. Given that, I can only really imagine two additional concerns with naively adding `/save-all` to the backup routine:

  1. Avoiding duplication: if online hosts already run `/save-all` every X minutes, running it with every backup duplicates work and I/O load. 
     
     **Fix:** This can pretty easily be fixed by check the last time `/save-all` was run, and only running it again if it hasn't run within the last backup period. E.g. if backups are scheduled to run every Y minutes, ensure that the latest call to `/save-all` was within the past Y minutes. If not, run `/save-all` before backing up. I imagine getting the last `/save-all` invocation time would be as simple as checking the latest mtime across all region files.

  2. Running to frequently: if online hosts run `/save-all` every X minutes to achieve some target load, if backups are configured to some different period Y minutes where Y < X additional load may be imposed on the server (even with the fix for 1. above).

     **Fix:** A second periodicity config argument could be added to the mod's config. Something like `save_all_period`. This can then be set to some default value that is sufficiently large to present minimal load (e.g. 30 or 60 minutes). Then when implementing the fix above for 1., you'd check if the last `/save-all` was more than `save_all_period` ago instead.

This could also simply be added behind a config flag like `save_all_on_backup` that defaults to false. Then users (like me) who would prefer this behavior can simply change the config to true.

### Short-term External Soln.

I agree that solving this on Linux is pretty easy, but I'm hosting the server on windows (since I run the client alongside it on the same machine). I'm not keen on wrapping the console with another program to pipe `/save-all` into the console periodically because then I lose all other console functionality. I could of course try to pipe through the console of the wrapper... but that feels very overengineered for this problem... and I have enough overengineering in my life rn ;)

### TL;DR
I guess my bug here boils down to: the average user running a modded server would expect that backups every X minutes means that in the event of a crash at most X minutes of playtime could be lost. As-is, this mod does not meet this criterion and therefore doesn't really function as a backup mod.

So far I have been unable to (usefully) recover from backups made by this mod 3 times, and have **never** once found an instance where these backups were useful (the region data is _always_ stale). Whereas with non-FTB packs I have been able to recover from backups generated by other mods multiple times without issue.

**A <10s backup time is not a useful feature if the resulting backups files cannot be used**.

**Comment by Raidobw2** [comment]
Thank you for the detailed reply. About the hotlist, I'm not fully sure what you mean, but here it's already a tracker for our main modder. In fact, our main modder is the one assigned to this issue, therefore he sees this. Of course, it depends on his current workload, which is always big and full of problems to solve, since he pretty much makes almost all FTB mods by himself and also participates in modpack development.

Basically, the way to resolve this ticket is one of two things:
- Give our main modder some time, I know this was opened already a few months ago and I thank you for your patience.
- We accept PRs if someone wants to code the solution.

I personally don't have the current knowledge to fix it, and am assigned to other tasks as well.

\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>\>
@LatvianModder Short summary: Subject of post is to improve FTB Backups' consistency when it does backups by having it flush to disk all data, including chunk data. Very unsure if current MC 1.16.5 autosaves on server. I think it requires a wrapper, but unsure again even after research.

The only indication of autosave being present in any Java MC version is in 1.18, this might be solved for the future? Thanks for seeing it Lat and have a nice day.
\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<
![image](https://user-images.githubusercontent.com/17010632/144698487-9a51a1aa-588e-451e-8318-60d9a19507b5.png)

**Comment by Raidobw2** [comment]
Mikey just as a very short TLDR of what this issue is:
- Backups should always flush to disk before backupping, includes region data and player inventories
- Autosave could be implemented into FTB Backups or FTB Essentials, and essentially force save a MC world every X minutes. That feature would only be valid for 1.16.5, because 1.18+ MC has built-in autosave afaik
